app/controllers/oaisys/pmh_controller.rb

```diff
@@ -144,13 +190,37 @@ def public_items_for_metadata_format(verb:, format:, page:, restricted_to_set: n
     [model, model.total_count, cursor]
   end
 
+  def sets_on_page(page:)
+    items_per_request = Oaisys::Engine.config.items_per_request
+    model = Oaisys::Engine.config.set_model
+
+    model = model.page(page).per(items_per_request)
+    cursor = (page - 1) * items_per_request
+    [model, model.total_count, cursor]
+  end
+
+  def expire_token(resumption_token:, verb:)
+    Oaisys::Engine.config.redis.expire_token(resumption_token: resumption_token, verb: verb,
+                                             identifier: user_identifier)
+  end
+
   def resumption_token_from_params(parameters:)
     parameters[:page] = parameters[:page] + 1
-    CGI.escape(parameters.except(:verb).to_query)
+    Oaisys::Engine.config.redis.create_token(parameters: parameters.except(:verb, :resumptionToken),
+                                             verb: parameters[:verb], identifier: user_identifier)
   end
 
-  def params_from_resumption_token(resumption_token:)
-    Rack::Utils.parse_query(CGI.unescape(resumption_token)).symbolize_keys
+  def params_from_resumption_token(resumption_token:, verb:)
+    Oaisys::Engine.config.redis.get_parameters(resumption_token: resumption_token, verb: verb,
+                                               identifier: user_identifier)
+  end
+
+  def user_identifier
```

Comment:
Probably better to call this `user_agent` as that's less likely to be misunderstood as an id for a user account.